crates/air-lib/air-parser/src/parser/validator.rs

```diff
@@ -30,6 +30,194 @@ use multimap::MultiMap;
 use std::collections::HashMap;
 use std::ops::Deref;
 
+#[derive(Clone, Copy, Debug)]
+pub enum CheckInstructionKind<'names> {
+    PivotalNext(&'names str),
+    Merging,
+    PopStack1,
+    PopStack2,
+    Replacing,
+    ReplacingWithCheck(&'names str),
+    Xoring,
+    PopStack1ReplacingWithCheck(&'names str),
+    Simple,
+}
+
+/// This machine is used to check that there are no instructions after next
+/// in a fold block over stream and map, e.g.
+/// (fold $iterable iterator
+///  (seq
+///    (next iterator)
+///    (call  ...)  <- instruction after next
+///  )
+/// )
+/// Note that the fold over scalar doesn't have this restriction.
+#[derive(Clone, Debug)]
+struct AfterNextCheckMachine<'name> {
+    /// stack for the machine.
+    stack: Vec<(CheckInstructionKind<'name>, Span)>,
+
+    /// lalrpop effectively parses from right to left thus `next`
+    /// happens before fold. It is impossible to tell whether `next`
+    /// belongs to a scalar fold or not.
+    potentially_impossible_spans: Vec<(&'name str, Span)>,
+
+    /// This vector contains all spans where an instruction after next was met.
+    impossible_spans: Vec<Span>,
```

Comment:
`denied_spans`, `violating_spans`, etc.  "Impossible" is too strong word.

Reply:
Renamed